In [63]:
import pandas as pd
import os

### Load data

In [83]:
def load_data(path, type):
    path = os.path.join(dir, path)
    with open(path) as f:
        lines = f.readlines()
        

    if type == "mm":
        data = lines[29:]
        data = [line.split() for line in data]
        df = pd.DataFrame(data, columns=["time", "vdw_protein", "elec_protein", "vdw_ligand", "elec_ligand", "vdw_complex", "elec_complex", "total_energy"]).astype(float)
    elif type == "polar":
        data = lines[25:]
        data = [line.split() for line in data]
        df = pd.DataFrame(data, columns =["time", "protein", "ligand", "complex"]).astype(float)
    elif type == "nonpolar":
        data = lines[31:]
        data = [line.split() for line in data]
        data = [line[:4] for line in data]
        df = pd.DataFrame(data, columns = ["time", "protein", "ligand", "complex"]).astype(float)
    else:
        raise ValueError("type must be either mm, polar, or nonpolar")
    
    return df

In [67]:
np = load_data("sasa.xvg", "nonpolar")

### Calculation

#### Protein

##### mm

In [205]:
mm_mean_protein = []

for i in range(len(mm)):
    mm_mean_protein.append(mm.iloc[i]["vdw_protein"] + mm.iloc[i]["elec_protein"])

##### polar

In [208]:
polar_protein =polar["protein"]

##### np

In [211]:
npolar_protein = npolar["protein"]

#### Ligand

##### mm

In [213]:
mm_mean_ligand = []

for i in range(len(mm)):
    mm_mean_ligand.append(mm.iloc[i]["vdw_ligand"] + mm.iloc[i]["elec_ligand"])


##### polar

In [216]:
polar_ligand = polar["ligand"]

##### np

In [218]:
npolar_ligand = npolar["ligand"]

#### complex

##### mm

In [221]:
mm_mean_complex = []

for i in range(len(mm)):
    mm_mean_complex.append(mm.iloc[i]["vdw_complex"] + mm.iloc[i]["elec_complex"])

##### polar

In [223]:
polar_complex = polar["complex"]

##### np

In [225]:
npolar_complex = npolar["complex"]

#### Data calculation

summary of all data calculation from calculation i.e mean of protein in mm dataset

In [227]:
df_calc = pd.DataFrame({"mm_mean_protein": mm_mean_protein, "mm_mean_ligand": mm_mean_ligand, "mm_mean_complex": mm_mean_complex, "polar_protein": polar_protein, "polar_ligand": polar_ligand, "polar_complex": polar_complex, "npolar_protein": npolar_protein, "npolar_ligand": npolar_ligand, "npolar_complex": npolar_complex})

In [228]:
df_calc.head()

mm_mean_protein  mm_mean_ligand  mm_mean_complex  polar_protein  \
0       -13523.615        -205.115       -13829.962     -13613.267   
1       -13112.673        -199.782       -13400.509     -13743.670   
2       -13177.128        -208.210       -13482.585     -13799.874   
3       -13467.088        -190.943       -13783.440     -13682.825   
4       -13521.661        -221.938       -13810.692     -13933.196   

   polar_ligand  polar_complex  npolar_protein  npolar_ligand  npolar_complex  
0      -154.148     -13676.761         394.645         14.559         399.328  
1      -150.056     -13809.009         402.692         15.277         408.334  
2      -141.300     -13860.200         404.829         13.837         410.009  
3      -143.622     -13702.220         403.211         14.176         408.128  
4      -145.774     -13925.234         414.127         14.544         418.121

### Calculate G

calculate G value for each parameter in each row dataset

In [234]:
df_calc["G_protein"] = df_calc["mm_mean_protein"] + df_calc["polar_protein"] + df_calc["npolar_protein"]
df_calc["G_ligand"] = df_calc["mm_mean_ligand"] + df_calc["polar_ligand"] + df_calc["npolar_ligand"]
df_calc["G_complex"] = df_calc["mm_mean_complex"] + df_calc["polar_complex"] + df_calc["npolar_complex"]


calculation result in dataset

In [235]:
df_calc.head()

mm_mean_protein  mm_mean_ligand  mm_mean_complex  polar_protein  \
0       -13523.615        -205.115       -13829.962     -13613.267   
1       -13112.673        -199.782       -13400.509     -13743.670   
2       -13177.128        -208.210       -13482.585     -13799.874   
3       -13467.088        -190.943       -13783.440     -13682.825   
4       -13521.661        -221.938       -13810.692     -13933.196   

   polar_ligand  polar_complex  npolar_protein  npolar_ligand  npolar_complex  \
0      -154.148     -13676.761         394.645         14.559         399.328   
1      -150.056     -13809.009         402.692         15.277         408.334   
2      -141.300     -13860.200         404.829         13.837         410.009   
3      -143.622     -13702.220         403.211         14.176         408.128   
4      -145.774     -13925.234         414.127         14.544         418.121   

   G_protein  G_ligand  G_complex  
0 -26742.237  -344.704 -27107.395  
1 -26453.651  -334.561 -26801.184  
2 -26572.173  -335.673 -26932.776  
3 -26746.702  -320.389 -27077.532  
4 -27040.730  -353.168 -27317.805

### Calculate Gbinding

calculate Gbindig for each row

In [236]:
df_calc["G_binding"] = df_calc["G_complex"] - df_calc["G_protein"] - df_calc["G_ligand"]

In [239]:
G_binding_mean = np.mean(df_calc["G_binding"])

In [240]:
G_binding_std = np.std(df_calc["G_binding"])

### df summary

calculation summary for each row

In [242]:
df_calc.head()

mm_mean_protein  mm_mean_ligand  mm_mean_complex  polar_protein  \
0       -13523.615        -205.115       -13829.962     -13613.267   
1       -13112.673        -199.782       -13400.509     -13743.670   
2       -13177.128        -208.210       -13482.585     -13799.874   
3       -13467.088        -190.943       -13783.440     -13682.825   
4       -13521.661        -221.938       -13810.692     -13933.196   

   polar_ligand  polar_complex  npolar_protein  npolar_ligand  npolar_complex  \
0      -154.148     -13676.761         394.645         14.559         399.328   
1      -150.056     -13809.009         402.692         15.277         408.334   
2      -141.300     -13860.200         404.829         13.837         410.009   
3      -143.622     -13702.220         403.211         14.176         408.128   
4      -145.774     -13925.234         414.127         14.544         418.121   

   G_protein  G_ligand  G_complex  G_binding  
0 -26742.237  -344.704 -27107.395    -20.454  
1 -26453.651  -334.561 -26801.184    -12.972  
2 -26572.173  -335.673 -26932.776    -24.930  
3 -26746.702  -320.389 -27077.532    -10.441  
4 -27040.730  -353.168 -27317.805     76.093

statistical summary from data calculated from G and G_binding

In [256]:
df_calc_describe = df_calc.describe()
df_calc_describe

mm_mean_protein  mm_mean_ligand  mm_mean_complex  polar_protein  \
count       201.000000      201.000000       201.000000     201.000000   
mean     -12917.699373     -202.891886    -13215.942433  -14444.830821   
std         274.924359       11.546392       280.908624     321.332992   
min      -13523.615000     -234.841000    -13829.962000  -15473.761000   
25%      -13107.106000     -209.754000    -13400.509000  -14656.937000   
50%      -12916.201000     -203.690000    -13232.279000  -14448.775000   
75%      -12741.485000     -195.605000    -13058.247000  -14241.930000   
max      -12157.553000     -171.804000    -12451.657000  -13613.267000   

       polar_ligand  polar_complex  npolar_protein  npolar_ligand  \
count    201.000000     201.000000      201.000000     201.000000   
mean    -149.452990  -14497.176856      421.118527      14.487891   
std        5.365655     326.797339        6.521346       0.517513   
min     -163.748000  -15470.877000      394.645000      12.848000   
25%     -152.948000  -14706.752000      417.112000      14.142000   
50%     -149.266000  -14503.888000      421.855000      14.513000   
75%     -145.894000  -14283.840000      425.284000      14.840000   
max     -136.597000  -13676.761000      435.987000      16.061000   

       npolar_complex     G_protein    G_ligand     G_complex   G_binding  
count      201.000000    201.000000  201.000000    201.000000  201.000000  
mean       426.700164 -26941.411667 -337.856985 -27286.419124   -7.150473  
std          6.655170    219.207166   12.043724    217.157477   43.680299  
min        399.328000 -27621.749000 -368.639000 -27870.547000 -104.331000  
25%        422.463000 -27065.539000 -346.584000 -27413.603000  -32.734000  
50%        427.204000 -26944.582000 -337.982000 -27293.954000  -16.248000  
75%        430.837000 -26800.056000 -329.389000 -27131.358000   13.929000  
max        441.613000 -26330.734000 -304.085000 -26656.456000  107.818000

### combine all dataframes into one (mm, polar, nonpolar)

In [258]:
df_all = pd.merge(mm, pd.merge(polar, npolar, on="time"), on="time")
df_all.head()

time  vdw_protein  elec_protein  vdw_ligand  elec_ligand  vdw_complex  \
0    0.0    -6257.871     -7265.744     102.666     -307.781    -6240.222   
1  100.0    -6028.126     -7084.547     106.574     -306.356    -5999.353   
2  200.0    -6105.704     -7071.424     102.497     -310.707    -6074.370   
3  300.0    -6402.911     -7064.177     118.080     -309.023    -6362.967   
4  400.0    -6403.868     -7117.793      92.539     -314.477    -6389.100   

   elec_complex  total_energy  protein_x  ligand_x  complex_x  protein_y  \
0     -7589.740      -101.232 -13613.267  -154.148 -13676.761    394.645   
1     -7401.156       -88.054 -13743.670  -150.056 -13809.009    402.692   
2     -7408.215       -97.248 -13799.874  -141.300 -13860.200    404.829   
3     -7420.473      -125.410 -13682.825  -143.622 -13702.220    403.211   
4     -7421.592       -67.093 -13933.196  -145.774 -13925.234    414.127   

   ligand_y  complex_y  
0    14.559    399.328  
1    15.277    408.334  
2    13.837    410.009  
3    14.176    408.128  
4    14.544    418.121

In [259]:
df_all.describe()

time  vdw_protein  elec_protein  vdw_ligand  elec_ligand  \
count    201.000000   201.000000    201.000000  201.000000   201.000000   
mean   10000.000000 -6136.975741  -6780.723632  102.782721  -305.674607   
std     5816.786054   140.468268    265.644694   10.902650     5.000400   
min        0.000000 -6590.913000  -7421.387000   76.172000  -318.005000   
25%     5000.000000 -6235.352000  -6940.355000   96.347000  -308.850000   
50%    10000.000000 -6141.737000  -6802.670000  102.624000  -305.716000   
75%    15000.000000 -6036.208000  -6614.167000  109.927000  -302.638000   
max    20000.000000 -5738.242000  -6058.880000  133.506000  -291.733000   

       vdw_complex  elec_complex  total_energy     protein_x    ligand_x  \
count   201.000000    201.000000    201.000000    201.000000  201.000000   
mean  -6103.052194  -7112.890239    -95.351189 -14444.830821 -149.452990   
std     137.599784    268.382804     26.587909    321.332992    5.365655   
min   -6558.427000  -7739.644000   -168.722000 -15473.761000 -163.748000   
25%   -6194.146000  -7273.540000   -113.470000 -14656.937000 -152.948000   
50%   -6103.016000  -7131.189000    -91.367000 -14448.775000 -149.266000   
75%   -6007.631000  -6940.503000    -74.897000 -14241.930000 -145.894000   
max   -5722.510000  -6373.048000    -36.129000 -13613.267000 -136.597000   

          complex_x   protein_y    ligand_y   complex_y  
count    201.000000  201.000000  201.000000  201.000000  
mean  -14497.176856  421.118527   14.487891  426.700164  
std      326.797339    6.521346    0.517513    6.655170  
min   -15470.877000  394.645000   12.848000  399.328000  
25%   -14706.752000  417.112000   14.142000  422.463000  
50%   -14503.888000  421.855000   14.513000  427.204000  
75%   -14283.840000  425.284000   14.840000  430.837000  
max   -13676.761000  435.987000   16.061000  441.613000

### Report

In [254]:
print ("### Summary ### \n")


print ("G_binding_mean: ", G_binding_mean)
print ("G_binding_std: ", G_binding_std)

print ("\n ### Summary ###")

### Summary ### 

G_binding_mean:  -7.150472636815841
G_binding_std:  43.57150642883602

 ### Summary ###


**the summary of the data is as follows:** 

* The mean of the binding free energy is -7.5 kcal/mol.
* The standard deviation of the binding free energy is 43.57.
* The std of protein, complex in mm and polar dataset is high because the scale of dataset also hight rather than the others like ligand and nonpolar dataset. you can check in the df_all as combine dataset of mm, polar, and nn into one. That is why the std of binding energy is also high (*note : check df.describe as statistical summary*)

